Add `copy_from` option to the Append processor

chrisberkhout opened the pull request:
### Summary

A `copy_from` option for the Append processor, matching the option in the Set processor.

This makes it possible to refer to existing non-String values.

### Discussion

The alternative is to use a Script processor, which requires Painless code to manually initialize the destination field and to avoid duplicates if necessary.

One variation on this use case is [here](https://github.com/elastic/integrations/blob/62aca1a5732c8413850695c9e9f27b5f814c194f/packages/trend_micro_vision_one/data_stream/telemetry/elasticsearch/ingest_pipeline/default.yml#L454-L467) and in the following 7 script processors.

I'd be happy to add an `ignore_empty_value` option as well,  which would make it easier to collect values that may exist in multiple locations into one destination.

---

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? ✅
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)? ✅
- If submitting code, have you built your formula locally prior to submission with `gradle check`? ✅
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed. ✅
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)? ✅
- ~If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.~

chrisberkhout committed:
Add copy_from option to append processor.

chrisberkhout committed:
Update documentation to add the copy_from option.

chrisberkhout committed:
Fix style.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @chrisberkhout, I've created a changelog YAML for you.

andrewkroh reviewed (commented):
Happy to have this feature to avoid templating.

This change will also need an accompanying modification to the elasticsearch-specification to add the new parameter at https://github.com/elastic/elasticsearch-specification/blob/e585438d116b00ff34643179e6286e402c0bcaaf/specification/ingest/_types/Processors.ts#L329-L344

chrisberkhout commented on docs/reference/enrich-processor/append-processor.md:
```diff
@@ -14,7 +14,8 @@ $$$append-options$$$
 | `field` | yes | - | The field to be appended to. Supports [template snippets](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#template-snippets). |
-| `value` | yes | - | The value to be appended. Supports [template snippets](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#template-snippets). |
+| `value` | yes* | - | The value to be appended. Supports [template snippets](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#template-snippets). May specify only one of `value` or `copy_from`. |
+| `copy_from` | no | - | The source field to be append. Cannot set `value` simultaneously. Supported data types are `boolean`, `number`, `array`, `object`, `string`, `date`, etc. |
```
Agreed. I was following the description in Set processor.

The limitation is that the (existing) [deep copy](https://github.com/elastic/elasticsearch/blob/14debad9bf0d1904180880e3cb2a52640353cc90/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L776-L820) will throw an `IllegalArgumentException` if it encounters something other than `Map`, `List`, `Set`, `byte[]`, `double[][]`, `double[]`, `null`, `String`, `Integer`, `Long`, `Float`, `Double`, `Boolean`, `ZonedDateTime`, `Date`.

I'll just remove the part about supported types.

chrisberkhout committed:
Docs: don't mention an incomplete list of supported types.

It mentioned 6 types, the full list is 15:

Map, List, Set, byte[], double[][], double[], null, String, Integer,
Long, Float, Double, Boolean, ZonedDateTime, Date.

chrisberkhout commented:
> This change will also need an accompanying modification to the elasticsearch-specification to add the new parameter at https://github.com/elastic/elasticsearch-specification/blob/e585438d116b00ff34643179e6286e402c0bcaaf/specification/ingest/_types/Processors.ts#L329-L344

I opened a draft PR for it: https://github.com/elastic/elasticsearch-specification/pull/5056

elasticsearchmachine commented:
Hi @chrisberkhout, I've created a changelog YAML for you.

chrisberkhout committed:
Update docs/changelog/132003.yaml

joegallo committed:
Fix a typo (missing a space)

joegallo committed:
Reword this slightly

To make it more similar to the equivalent docs from the `set`
processor.

joegallo committed:
Add an applies_to tag for the copy_from field

joegallo committed:
Add factory unit tests

joegallo committed:
Ensure the values are unique for this test

joegallo committed:
Add a simple test

joegallo commented:
> I'd be happy to add an `ignore_empty_value` option as well, which would make it easier to collect values that may exist in multiple locations into one destination.

There's a PR up for that one already (which your comment reminded me of). My intention is to get this one merged and then return to that one so that they're both in for 9.2.0.

chrisberkhout commented:
@joegallo Yeah, sorry about that. I used Github's "Update with rebase" button, so no manual changes but I get your point.

joegallo commented:
No worries! It's a small thing.

joegallo committed:
Merge branch 'main' into append-processor-copy-from

joegallo committed:
Merge branch 'main' into append-processor-copy-from

joegallo committed:
Add rest tests

joegallo committed:
Merge branch 'main' into append-processor-copy-from

andrewkroh commented on docs/reference/enrich-processor/append-processor.md:
```diff
@@ -14,9 +14,10 @@ $$$append-options$$$
+| `media_type` | no | `application/json` | The media type for encoding `value`. Applies only when `value` is a [template snippet](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#template-snippets). Must be one of `application/json`, `text/plain`, or`application/x-www-form-urlencoded`. |
 | `description` | no | - | Description of the processor. Useful for describing the purpose of the processor or its configuration. |
 | `if` | no | - | Conditionally execute the processor. See [Conditionally run a processor](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#conditionally-run-processor). |
 | `ignore_failure` | no | `false` | Ignore failures for the processor. See [Handling pipeline failures](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#handling-pipeline-failures). |
```
Consider adding an example below to encourage `copy_from` usage. Since it doesn't have an `ignore_empty_value`, then maybe with an `if` to demonstrate how to accomplish that...

```json
{
  "append": {
    "if": "ctx.host?.name instanceof String && !ctx.host.name.isEmpty()",
    "field": "related.hosts",
    "copy_from": "host.name",
    "allow_duplicates": false
  }
}
```

joegallo replied:
https://www.elastic.co/docs/reference/enrich-processor/append-processor#append-processor-example-using-allow-duplicates-and-ignore-empty-values
